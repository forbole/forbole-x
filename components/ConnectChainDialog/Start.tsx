import {
  Table,
  TableBody,
  TableHead,
  DialogContent,
  TableRow,
  TableCell,
  Typography,
  Button,
  Box,
  Avatar,
  useTheme,
} from '@material-ui/core'
import useTranslation from 'next-translate/useTranslation'
import React from 'react'
import format from 'date-fns/format'
import LogoLight from '../../assets/images/image_default_light.svg'
import LogoDark from '../../assets/images/image_default_dark.svg'
import ConnectIcon from '../../assets/images/icons/icon_connect.svg'
import useStyles from './styles'
import connectableChains from '../../misc/connectableChains'
import { useGeneralContext } from '../../contexts/GeneralContext'
import useIconProps from '../../misc/useIconProps'
import useSendTransaction from '../../misc/tx/useSendTransaction'
import { useWalletsContext } from '../../contexts/WalletsContext'

interface StartProps {
  account: Account
  onConnectClick(): void
  onClose(): void
  connections: ChainConnection[]
}

const Start: React.FC<StartProps> = ({ onConnectClick, connections, account, onClose }) => {
  const { t } = useTranslation('common')
  const classes = useStyles()
  const { theme } = useGeneralContext()
  const themeStyle = useTheme()
  const sendTransaction = useSendTransaction()
  const { password } = useWalletsContext()
  const iconProps = useIconProps(undefined, themeStyle.palette.primary.main)

  return (
    <DialogContent className={classes.dialogContent}>
      {connections.length ? (
        <>
          <Table className={classes.borderedTable}>
            <TableHead>
              <TableRow>
                <TableCell className={classes.tableCell}>{t('chain account')}</TableCell>
                <TableCell className={classes.tableCell}>{t('creation time')}</TableCell>
                <TableCell align="center" className={classes.tableCell}>
                  {t('manage')}
                </TableCell>
              </TableRow>
            </TableHead>

            <TableBody>
              {connections.map(({ creationTime, externalAddress, chainName }, i) => {
                const chain = connectableChains[chainName]
                return (
                  <TableRow key={chainName + externalAddress} className={classes.borderedTableRow}>
                    <TableCell className={classes.tableCell}>
                      <Box display="flex">
                        <Avatar
                          src={chain ? chain.image : undefined}
                          alt={chain ? chain.name : chainName}
                        />
                        <Box ml={1}>
                          <Typography>{chain ? chain.name : chainName}</Typography>
                          <Typography variant="body2" color="textSecondary">
                            {externalAddress}
                          </Typography>
                        </Box>
                      </Box>
                    </TableCell>
                    <TableCell className={classes.tableCell}>
                      {format(creationTime, 'dd MMM yyyy, HH:mm:ss')}
                    </TableCell>
                    <TableCell align="center" className={classes.tableCell}>
                      <Button
                        color="primary"
                        onClick={async () => {
                          await sendTransaction(password, account.address, {
                            msgs: [
                              {
                                typeUrl: '/desmos.profiles.v3.MsgUnlinkChainAccount',
                                value: {
                                  chainName,
                                  owner: account.address,
                                  target: externalAddress,
                                },
                              },
                            ],
                            memo: '',
                          })
                          onClose()
                        }}
                      >
                        {t('disconnect')}
                      </Button>
                    </TableCell>
                  </TableRow>
                )
              })}
            </TableBody>
          </Table>
          <Box m={2} display="flex" justifyContent="flex-end">
            <Button
              startIcon={<ConnectIcon {...iconProps} />}
              onClick={onConnectClick}
              color="primary"
            >
              {t('connect chain')}
            </Button>
          </Box>
        </>
      ) : (
        <Box display="flex" flexDirection="column" alignItems="center" mt={2} mb={6}>
          {theme === 'dark' ? (
            <LogoDark width={themeStyle.spacing(30)} />
          ) : (
            <LogoLight width={themeStyle.spacing(30)} />
          )}
          <Box mt={3} mb={2}>
            <Typography>{t('connect chain description')}</Typography>
          </Box>

          <Button onClick={onConnectClick} variant="contained" color="primary">
            {t('connect chain')}
          </Button>
        </Box>
      )}
    </DialogContent>
  )
}

export default Start
