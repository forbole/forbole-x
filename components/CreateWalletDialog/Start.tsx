import {
  ButtonBase,
  Button,
  DialogActions,
  DialogContent,
  DialogContentText,
  Grid,
  Typography,
  Box,
} from '@material-ui/core'
import useTranslation from 'next-translate/useTranslation'
import React from 'react'
import WithMnemonicIconLight from '../../assets/images/login_light.svg'
import WithMnemonicIconDark from '../../assets/images/login_dark.svg'
import WithoutMnemonicIconLight from '../../assets/images/create_wallet_light.svg'
import WithoutMnemonicIconDark from '../../assets/images/create_wallet_dark.svg'
import useStyles from './styles'
import { useGeneralContext } from '../../contexts/GeneralContext'

interface StartProps {
  onImportWalletClick(): void
  onCreateWalletClick(): void
  onWhatIsMnemonicClick(): void
}

const Start: React.FC<StartProps> = ({
  onImportWalletClick,
  onCreateWalletClick,
  onWhatIsMnemonicClick,
}) => {
  const { t } = useTranslation('common')
  const classes = useStyles()
  const { theme } = useGeneralContext()

  return (
    <>
      <DialogContent className={classes.dialogContent}>
        <DialogContentText>{t('create wallet description')}</DialogContentText>
        <Grid container spacing={5}>
          <Grid item xs={6}>
            <ButtonBase className={classes.selectionBox} onClick={onImportWalletClick}>
              <Box mb={10}>
                {theme === 'light' ? <WithMnemonicIconLight /> : <WithMnemonicIconDark />}
              </Box>
              <Typography variant="body2" align="center" color="textSecondary">
                {t('i have secret recovery phrase')}
              </Typography>
            </ButtonBase>
          </Grid>
          <Grid item xs={6}>
            <ButtonBase className={classes.selectionBox} onClick={onCreateWalletClick}>
              <Box mb={10}>
                {theme === 'light' ? <WithoutMnemonicIconLight /> : <WithoutMnemonicIconDark />}
              </Box>
              <Typography variant="body2" align="center" color="textSecondary">
                {t('i dont have secret recovery phrase')}
              </Typography>
            </ButtonBase>
          </Grid>
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button className={classes.button} color="primary" onClick={onWhatIsMnemonicClick}>
          {t('what is secret recovery phrase')}
        </Button>
      </DialogActions>
    </>
  )
}

export default Start
