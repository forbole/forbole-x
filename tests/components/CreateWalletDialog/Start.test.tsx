import React from 'react'
import renderer from 'react-test-renderer'
import Start from '../../../components/CreateWalletDialog/Start'

const onImportWalletClick = jest.fn()
const onCreateWalletClick = jest.fn()
const onWhatIsMnemonicClick = jest.fn()

describe('component: CreateWalletDialog - Start', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <Start
        onWhatIsMnemonicClick={onWhatIsMnemonicClick}
        onImportWalletClick={onImportWalletClick}
        onCreateWalletClick={onCreateWalletClick}
      />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('calls onImportWalletClick when first button is clicked', () => {
    const component = renderer.create(
      <Start
        onWhatIsMnemonicClick={onWhatIsMnemonicClick}
        onImportWalletClick={onImportWalletClick}
        onCreateWalletClick={onCreateWalletClick}
      />
    )
    renderer.act(() => {
      component.root.findAllByType('button')[0].props.onClick()
    })
    expect(onImportWalletClick).toBeCalled()
  })
  it('calls onCreateWalletClick when second button is clicked', () => {
    const component = renderer.create(
      <Start
        onWhatIsMnemonicClick={onWhatIsMnemonicClick}
        onImportWalletClick={onImportWalletClick}
        onCreateWalletClick={onCreateWalletClick}
      />
    )
    renderer.act(() => {
      component.root.findAllByType('button')[1].props.onClick()
    })
    expect(onCreateWalletClick).toBeCalled()
  })
  it('calls onWhatIsMnemonicClick when last button is clicked', () => {
    const component = renderer.create(
      <Start
        onWhatIsMnemonicClick={onWhatIsMnemonicClick}
        onImportWalletClick={onImportWalletClick}
        onCreateWalletClick={onCreateWalletClick}
      />
    )
    renderer.act(() => {
      component.root.findAllByType('button')[2].props.onClick()
    })
    expect(onWhatIsMnemonicClick).toBeCalled()
  })
})

afterEach(() => {
  jest.clearAllMocks()
})
