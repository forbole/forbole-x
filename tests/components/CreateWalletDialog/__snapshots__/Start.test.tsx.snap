// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: CreateWalletDialog - Start renders correctly 1`] = `
Array [
  <div
    className="MuiDialogContent-root HookGlobalStyles-dialogContent-7"
  >
    <p
      className="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      common:create wallet description
    </p>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        <button
          className="MuiButtonBase-root HookGlobalStyles-selectionBox-4"
          disabled={false}
          onBlur={[Function]}
          onClick={[MockFunction]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <div
            className="MuiBox-root MuiBox-root-8"
          >
            <svg
              data-file-name="SvgLogin_light"
            />
          </div>
          <p
            className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
          >
            common:i have mnemonic phrase
          </p>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        <button
          className="MuiButtonBase-root HookGlobalStyles-selectionBox-4"
          disabled={false}
          onBlur={[Function]}
          onClick={[MockFunction]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <div
            className="MuiBox-root MuiBox-root-9"
          >
            <svg
              data-file-name="SvgCreate_wallet_light"
            />
          </div>
          <p
            className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
          >
            common:i dont have mnemonic phrase
          </p>
        </button>
      </div>
    </div>
  </div>,
  <div
    className="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text HookGlobalStyles-button-1 MuiButton-textSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        common:what is mnemonic phrase
      </span>
    </button>
  </div>,
]
`;
